jsettlers.graphics/src/main/java/jsettlers/graphics/map/draw/MapObjectDrawer.java

```diff
@@ -180,6 +192,174 @@ public void drawMapObject(int x, int y, IMapObject object) {
+		final int imageFile = 36;
+		ImageLink shipLink;
+		int sequence;
+		Image image;
+		if (state < .95) {
+			// draw ship under construction
+			sequence = (shipType == EMovableType.FERRY) ? 7 : 3;
+			shipLink = new OriginalImageLink(EImageLinkType.SETTLER, imageFile, sequence, shipDirection.ordinal);
+			image = imageProvider.getImage(shipLink);
+			drawWithConstructionMask(x, y, state, image, shade);
+		} else {
+			// get drawing position
+			Color color = context.getPlayerColor(ship.getPlayer().getPlayerId());
+			float viewX;
+			float viewY;
+			if (ship.getAction() == EMovableAction.WALKING) {
+				int originX = x - direction.getGridDeltaX();
+				int originY = y - direction.getGridDeltaY();
+				viewX = betweenTilesX(originX, originY, x, y, ship.getMoveProgress());
+				viewY = betweenTilesY;
+			} else {
+				viewX = mc.getViewX(x, y, 0);
+				viewY = mc.getViewY(x, y, 0);
+			}
+			// draw ship body
+			sequence = (shipType == EMovableType.FERRY) ? 4 : 0;
+			shipLink = new OriginalImageLink(EImageLinkType.SETTLER, imageFile, sequence, shipDirection.ordinal);
+			image = imageProvider.getImage(shipLink);
+			image.drawAt(gl, db, viewX, viewY, color, shade);
+			// prepare freight drawing
+			ArrayList<IMovable> passengerList = ship.getPassengers();
+			byte[] dx = EDirection.getXDeltaArray();
+			byte[] dy = EDirection.getYDeltaArray();
+			float baseViewX = mc.getViewX(x, y, 0);
+			float baseViewY = mc.getViewY(x, y, 0);
+			float xShiftForward = mc.getViewX(x + dx[direction.ordinal], y + dy[direction.ordinal], 0) - baseViewX;
+			float yShiftForward = mc.getViewY(x + dx[direction.ordinal], y + dy[direction.ordinal], 0) - baseViewY;
+			int xRight = x + dx[direction.rotateRight(1).ordinal] + dx[direction.rotateRight(2).ordinal];
+			int yRight = y + dy[direction.rotateRight(1).ordinal] + dy[direction.rotateRight(2).ordinal];
+			float xShiftRight = (mc.getViewX(xRight, yRight, 0) - baseViewX) / 2;
+			float yShiftRight = (mc.getViewY(xRight, yRight, 0) - baseViewY) / 2;
+			ArrayList<Float> yPosition = new ArrayList<>();
+			ArrayList<Integer> drawSequence = new ArrayList<>();
+			int numberOfFreight;
+			// get freight positions
+			if (shipType == EMovableType.FERRY) {
+				numberOfFreight = passengerList.size();
+				if (numberOfFreight > maxNumberOfPassengers) {
+					numberOfFreight = maxNumberOfPassengers;
+				}
+				for (int i = 0; i < numberOfFreight; i++) {
+					yPosition.add(PASSENGER_POSITION_TO_FRONT[i] * yShiftForward + PASSENGER_POSITION_TO_RIGHT[i] * yShiftRight);
+					drawSequence.add(i);
+				}
+			} else {
+				numberOfFreight = ship.getNumberOfStacks();
+				if (numberOfFreight > maxNumberOfStacks) {
+					numberOfFreight = maxNumberOfStacks;
+				}
+				for (int i = 0; i < numberOfFreight; i++) {
+					yPosition.add(CARGO_POSITION_TO_FRONT[i] * yShiftForward + CARGO_POSITION_TO_RIGHT[i] * yShiftRight);
+					drawSequence.add(i);
+				}
+			}
+			// sort freight by view y
```

Comment:
Please use the utilities provided by Java for this.

I think you might have tried to avoid creating small objects during drawing. But `yPosition.add` implicitly creates a float object. Instead of creating a float, you can create your own object that is `Comparable`.